NiL.JS/BaseLibrary/Console.cs

```diff
@@ -47,22 +165,209 @@ public static void asserta(Function f, JSValue sample)
+            return JSValue.undefined;
+        }
+
+        public JSValue groupEnd(JSValue thisBind, Arguments args)
+        {
+            if (_groups.Count == 0)
+                _groups.RemoveAt(_groups.Count - 1);
+            
+            return JSValue.undefined;
+        }
+
+
+        public JSValue time(JSValue thisBind, Arguments args)
+        {
+            string label = "";
+            if (args.length > 0)
+                label = (args[0] ?? "null").ToString();
+
+            if (_timers.ContainsKey(label))
+                _timers[label].Restart();
+            else
+                _timers.Add(label, Stopwatch.StartNew());
+
+            return JSValue.undefined;
+        }
+
+        public JSValue timeEnd(JSValue thisBind, Arguments args)
+        {
+            string label = "";
+            if (args.length > 0)
+                label = (args[0] ?? "null").ToString();
+
+            double elapsed = 0.0;
+            if (_timers.ContainsKey(label))
+            {
+                _timers[label].Stop();
+                elapsed = (double)_timers[label].ElapsedTicks / Stopwatch.Frequency * 1000.0;
+                _timers.Remove(label);
+            }
+            
+            if (label != "")
+                label += ": ";
+            LogMessage(LogLevel.Info, label + Tools.DoubleToString(System.Math.Round(elapsed, 10)) + "ms");
+
+            return JSValue.undefined;
+        }
+
+        
+
+        internal static List<JSConsole> _consoles;
+        internal static void AddConsole(JSConsole console)
+        {
+            if (_consoles == null)
+                _consoles = new List<JSConsole>(4);
+            _consoles.Add(console);
+        }
+        internal static void RemoveConsole(JSConsole console)
+        {
+            if (_consoles == null)
+                _consoles = new List<JSConsole>(4);
+            _consoles.Add(console);
+        }
+        public static JSConsole GetConsole(GlobalContext globalContext)
+        {
+            if (_consoles == null)
+                return null;
+            for (int i = 0; i < _consoles.Count; i++)
             {
-                if (i > 0)
-                    System.Console.Error.Write(" ");
-                var r = args[i].ToString();
-                System.Console.Error.Write(r);
+                if (_consoles[i]._globalContext == globalContext)
+                    return _consoles[i];
             }
-            System.Console.Error.WriteLine();
             return null;
         }
+
+        public static JSValue CreateConsoleObject(GlobalContext globalContext)
+        {
+            return CreateConsoleObject(CreateConsole(globalContext));
+        }
+        public static JSValue CreateConsoleObject(JSConsole console)
+        {
+            return console.ConsoleObject;
+        }
+
+        public static JSConsole CreateConsole(GlobalContext globalContext)
+        {
+            JSConsole c = GetConsole(globalContext);
```

Comment:
Yeah, sure, it will be. I want to get rid of the code handling the 1 JSConsole for 1 GlobalContext all together but I will wait until I found an elegant solution.
If possible I want to store the current JSConsole for a given GlobalContext, so that the user can access and modify it. This was the whole purpose of this structure. It's only a very unpleasant side effect that it disables you from creating new JSConsoles...